File: LeetCode/Python Solutions/Shortest Path Visiting All Nodes/shortest.py
from collections import deque

class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        n = len(graph)
        target = (1 << n) - 1 # This is when all nodes are visited
        visited = set()
        queue = deque()
        
        # Initialize by adding all nodes to the queue
        for i in range(n):
            queue.append((i, 1 << i))  # (node, state)
            visited.add((i, 1 << i))
        
        step = 0
        while queue:
            size = len(queue)
            for i in range(size):
                node, state = queue.popleft()
                if state == target:
                    return step
                
                for next_node in graph[node]:
                    next_state = state | (1 << next_node)
                    if (next_node, next_state) not in visited:
                        visited.add((next_node, next_state))
                        queue.append((next_node, next_state))
            
            step += 1
        
        return -1

File: LeetCode/Java Solutions/Shortest Path Visiting All Nodes/shortest.java
import java.util.*;

class Solution {
    public int shortestPathLength(int[][] graph) {
        int n = graph.length;
        int target = (1 << n) - 1;
        Set<String> visited = new HashSet<>();
        Queue<int[]> queue = new LinkedList<>();
        
        // Initialize by adding all nodes to the queue
        for (int i = 0; i < n; i++) {
            queue.offer(new int[]{i, 1 << i});
            visited.add(i + "," + (1 << i));
        }
        
        int step = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] current = queue.poll();
                int node = current[0];
                int state = current[1];
                if (state == target) {
                    return step;
                }
                
                for (int nextNode : graph[node]) {
                    int nextState = state | (1 << nextNode);
                    String key = nextNode + "," + nextState;
                    if (!visited.contains(key)) {
                        visited.add(key);
                        queue.offer(new int[]{nextNode, nextState});
                    }
                }
            }
            step++;
        }
        
        return -1;
    }
}
